cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -148,9 +161,16 @@ public RunId apply(RunRecordMeta runRecordMeta) {
     };
 
   public AppMetadataStore(Table table, CConfiguration cConf, AtomicBoolean upgradeComplete) {
+    this(table, cConf, upgradeComplete, new NoOpMessagePublisher());
```

Comment:
What is the NoOp one is for??

Reply:
This kind of circular dependency seems not good. Why not having the `AppMetadataStore` methods return something so that the caller (the program state notification subscriber) can use it to determine what to do?